### PR TITLE
Replace validate() with onInputChange()

### PR DESCRIPTION
**Pull request by prybicki**

### CODE REVIEW

File: src/graph/SpatialMergePointsNode.cpp [hunk 1]
```diff
@@ -31,12 +31,13 @@ void SpatialMergePointsNode::setParameters(const std::vector<rgl_field_t>& field
...
 
-void SpatialMergePointsNode::validate()
+void SpatialMergePointsNode::onInputChange()
 {
 	pointInputs = Node::filter<IPointsNode>(this->inputs);
 
-	if (pointInputs.empty()) {
-		auto msg = "expected at least one IPointsNode input for SpatialMergePointsNode";
+	if (pointInputs.size() < 2) {
+		// Having just one input does not make sense, because it is equivalent to removing this Node.
```
Comment (PiotrMrozik) [inline src/graph/SpatialMergePointsNode.cpp:39]: I would stop right there for a while.
In the case of single input, this node is just a pass-through. This is not a mistake, just useless node.
I can imagine situation when we create a graph with spatial merge inside, only one input. This graph is working for some time, and when the situation is ready, we add second input (node or subgraph) to an already created and working graph.
This is not a no-go, however, re-think it.

**Review by PiotrMrozik — APPROVED** [review]
I like it.
Only few small questions.

File: src/graph/SpatialMergePointsNode.cpp (hunk 1 above)
Comment (prybicki) [reply src/graph/SpatialMergePointsNode.cpp:39]: So, why not add SpatialMerge when it's actually needed? Having it operate on just 1 input smells very much like a bug  / unintended situation.

Comment (prybicki) [reply src/graph/SpatialMergePointsNode.cpp:39]: So, why not add SpatialMerge when it's actually needed? Having it operate on just 1 input smells very much like a bug  / unintended situation.

File: src/graph/SpatialMergePointsNode.cpp [hunk 2]
```diff
@@ -48,10 +49,9 @@ void SpatialMergePointsNode::validate()
 		}
 
 		// Check input pointcloud has required fields
-		for (const auto& requiredField : std::views::keys(mergedData)) {
+		for (const auto& requiredField : getRequiredFieldList()) {
```
[the remark this answers is not in the document]
Comment (prybicki) [reply src/graph/SpatialMergePointsNode.cpp:52]: SSOT

File: src/macros/cuda.hpp [hunk 3]
```diff
@@ -36,6 +36,16 @@ do
 }                                                                                                   \
 while(false)
 
+#define CHECK_CUDA_NO_THROW(call) \
```
[the remark this answers is not in the document]
Comment (prybicki) [reply src/macros/cuda.hpp:39]: I have been bitten once again by throwing destructor and this time I didn't tolerate it.
I will ask @nebraszka to inspect all other destructors, good task for a fresh Junior :)

File: src/graph/Interfaces.hpp [hunk 4]
```diff
@@ -78,33 +86,44 @@ struct IPointsNode
...
+	virtual VArray::ConstPtr getFieldData(rgl_field_t field, cudaStream_t stream) const override
 	{ return input->getFieldData(field, stream); }
 
 protected:
-	IPointsNode::Ptr input;
+	IPointsNode::Ptr input {0};
 };
 
-struct IPointsNodeMultiInput : IPointsNode
+struct INoInputNode : virtual Node
```
[the remark this answers is not in the document]
Comment (prybicki) [reply src/graph/Interfaces.hpp:120]: Hahaha, yeah, that's funny.

File: src/graph/SpatialMergePointsNode.cpp (hunk 1 above)
Comment (PiotrMrozik) [reply src/graph/SpatialMergePointsNode.cpp:39]: I partially agree with You.
I would warn the user about it, but still let it.

Comment (prybicki) [reply src/graph/SpatialMergePointsNode.cpp:39]: I get your reasoning; also, consider the fact that the warning would land in log file and / or stdout, which requires additional effort to see in a typical RGL usage scenario (engine plugins). Therefore, I prefer an error, which is a minor inconvenience if the passthrough was desired and a huge helper if it was not.

**Pull request merged**